File: src/main/java/com/enonic/explorer/Locales.java
package com.enonic.explorer;

import java.util.Arrays;
import java.util.Locale;

import static com.google.common.base.Strings.isNullOrEmpty;
import static com.google.common.base.Strings.nullToEmpty;

import com.enonic.xp.script.bean.BeanContext;
import com.enonic.xp.script.bean.ScriptBean;


public final class Locales implements ScriptBean
{
	private String getFormattedDisplayName( Locale locale )
    {
        return locale.getDisplayName( locale ) + " (" + locale.toLanguageTag() + ")";
    }

	public Locale[] getLocales( final String query )
    {
        Locale[] locales = Locale.getAvailableLocales();
        if ( !nullToEmpty( query ).isBlank() )
        {
            String trimmedQuery = query.trim().toLowerCase();
            locales = Arrays.stream( locales ).
                filter( locale -> nullToEmpty( locale.toLanguageTag() ).toLowerCase().contains( trimmedQuery ) ||
                    nullToEmpty( locale.getDisplayName( locale ) ).toLowerCase().contains( trimmedQuery ) ||
                    nullToEmpty( locale.getLanguage() ).toLowerCase().contains( trimmedQuery ) ||
                    nullToEmpty( locale.getDisplayLanguage( locale ) ).toLowerCase().contains( trimmedQuery ) ||
                    nullToEmpty( locale.getVariant() ).toLowerCase().contains( trimmedQuery ) ||
                    nullToEmpty( locale.getDisplayVariant( locale ) ).toLowerCase().contains( trimmedQuery ) ||
                    nullToEmpty( locale.getCountry() ).toLowerCase().contains( trimmedQuery ) ||
                    nullToEmpty( locale.getDisplayCountry( locale ) ).toLowerCase().contains( trimmedQuery ) ||
                    nullToEmpty( getFormattedDisplayName( locale ) ).toLowerCase().contains( trimmedQuery ) &&
                        !isNullOrEmpty( locale.toLanguageTag() ) && !isNullOrEmpty( locale.getDisplayName() ) ).
                toArray( Locale[]::new );
        }
        else
        {
            locales = Arrays.stream( locales ).
                filter( ( locale ) -> !isNullOrEmpty( locale.toLanguageTag() ) && !isNullOrEmpty( locale.getDisplayName() ) ).
                toArray( Locale[]::new );
        }
		return locales;
    }

	@Override
    public void initialize( final BeanContext context )
    {
        //no-op
    }
}

File: src/main/resources/lib/locales.ts
type Locale = {
	getCountry :() => string
	getDisplayCountry :(locale :Locale) => string
	getDisplayLanguage :(locale :Locale) => string
	getDisplayName :(locale :Locale) => string
	getDisplayVariant :(locale :Locale) => string
	getLanguage :() => string
	toLanguageTag :() => string
	getVariant :() => string
};

type JavaLocale = unknown;

declare const __ :{
	newBean :<T>(s :string) => T
	toNativeObject :<T>(t :T) => T
}

declare const Java :{
	from :(javaLocales :Array<JavaLocale>) => Array<Locale>,
	type :<T>(s :string) => T
};


const {forLanguageTag} = Java.type<
	{forLanguageTag :(locale :string) => Locale}
		>('java.util.Locale');


export const getLocales = ({
	locale = undefined,
	query = ''
} : {
	locale ?:string
	query ?:string
} = {}) => {
	const uniqTagObj = {};
	Java.from(
		__.newBean<
			{getLocales :(query :string) => Array<JavaLocale>}
				>('com.enonic.explorer.Locales').getLocales(query)
	).forEach((l) => {
		const tag = l.toLanguageTag();
		if (!uniqTagObj[tag]) {
			uniqTagObj[tag] = {
				country: l.getCountry(),
				displayCountry: l.getDisplayCountry(locale ? forLanguageTag(locale) : l),
				displayLanguage: l.getDisplayLanguage(locale ? forLanguageTag(locale) : l),
				displayName: l.getDisplayName(locale ? forLanguageTag(locale) : l), // not null!
				displayVariant: l.getDisplayVariant(locale ? forLanguageTag(locale) : l),
				language: l.getLanguage(), // not null! uniq?
				tag, // uniq & not null!
				variant: l.getVariant()
			};
		}
	});
	return Object.keys(uniqTagObj).sort().map((k) => uniqTagObj[k]);
};
